clippy_lints/src/swap.rs

```diff
@@ -70,9 +71,133 @@ impl<'tcx> LateLintPass<'tcx> for Swap {
     fn check_block(&mut self, cx: &LateContext<'tcx>, block: &'tcx Block<'_>) {
         check_manual_swap(cx, block);
         check_suspicious_swap(cx, block);
+        check_xor_swap(cx, block);
     }
 }
 
+fn root_of_borrow<'a, 'hir>(e: &'a Expr<'hir>) -> (&'a Expr<'hir>, u32) {
+    if let ExprKind::Index(ep, _) = e.kind {
+        let (root, depth) = root_of_borrow(ep);
+        (root, depth + 1)
+    } else if let ExprKind::Field(ep, _) = e.kind {
+        let (root, depth) = root_of_borrow(ep);
+        (root, depth + 1)
+    } else {
+        (e, 0)
+    }
+}
+
+fn go_up<'a, 'hir>(e: &'a Expr<'hir>, depth: u32) -> &'a Expr<'hir> {
+    if depth == 0 {
+        e
+    } else if let ExprKind::Index(ep, _) = e.kind {
+        go_up(ep, depth - 1)
+    } else if let ExprKind::Field(ep, _) = e.kind {
+        go_up(ep, depth - 1)
+    } else {
+        unreachable!("depth is too much");
+    }
+}
+
+fn is_swappable(cx: &LateContext<'_>, mut e1: &Expr<'_>, mut e2: &Expr<'_>) -> bool {
+    let (r1, d1) = root_of_borrow(e1);
+    let (r2, d2) = root_of_borrow(e2);
+    if !eq_expr_value(cx, r1, r2) {
+        return true;
+    }
+    if d1 > d2 {
+        e1 = go_up(e1, d1 - d2);
+    } else {
+        e2 = go_up(e2, d2 - d1);
+    }
+    loop {
+        let (ep1, ep2) = match (&e1.kind, &e2.kind) {
+            (ExprKind::Index(ep1, _), ExprKind::Index(ep2, _)) => {
+                if eq_expr_value(cx, ep1, ep2) {
+                    return false;
+                }
+                (ep1, ep2)
+            },
+            (ExprKind::Field(ep1, _), ExprKind::Field(ep2, _)) => {
+                if eq_expr_value(cx, ep1, ep2) {
+                    return true;
+                }
+                (ep1, ep2)
+            },
+            (ExprKind::Index(ep1, _), ExprKind::Field(ep2, _)) | (ExprKind::Field(ep1, _), ExprKind::Index(ep2, _)) => {
+                (ep1, ep2)
+            },
+            _ => return false,
+        };
+        e1 = ep1;
+        e2 = ep2;
+    }
```

Comment:
There are a lot of redundant HIR traversals here: `root_of_borrow`, each `eq_expr_value` call, each`go_up` call, and finally this loop. You can traverse the HIR just once for `e1` and `e2` and push `Expr`s onto a stack.